docs/compilation/performance_tips.md

```diff
@@ -0,0 +1,645 @@
+# Performance tips
+
+## Improve parallelism
+
+Modern CPUs have multiple cores to perform computation and utilizing multiple cores is a great way to boost performance.
+
+There are two kinds of parallelism in Concrete:
+- Loop parallelism to make tensor operations parallel, achieved by using [OpenMP](https://www.openmp.org/)
+- Dataflow parallelism to make independent operations parallel, achieved by using [HPX](https://hpx.stellar-group.org/)
+
+Loop parallelism is enabled by default, as it's supported on all platforms. Dataflow parallelism however is only supported on Linux, hence not enabled by default.
+
+Here are some ways to improve parallelism.
+
+### By enabling dataflow parallelism
+
+Dataflow parallelism is a great feature, especially when the circuit is doing a lot of scalar operations.
+
+Without dataflow parallelism, circuit is executed operation by operation, like an imperative language. If the operations themselves are not tensorized, loop parallelism would not be utilized and the entire execution would happen in a single thread. Dataflow parallelism changes this by analyzing the operations and their dependencies within the circuit to determine what can be done in parallel and what cannot. Then it distributes the tasks that can be done in parallel to different threads.
+
+For example:
+
+```python
+import time
+
+import numpy as np
+from concrete import fhe
+
+def f(x, y, z):
+    # normally, you'd use fhe.array to construct a concrete tensor
+    # but for this example, we just create a simple numpy array
+    # so the matrix multiplication can happen on a cellular level
+    a = np.array([[x, y], [z, 2]])
+    b = np.array([[1, x], [z, y]])
+    return fhe.array(a @ b)
+
+inputset = fhe.inputset(fhe.uint3, fhe.uint3, fhe.uint3)
+
+for dataflow_parallelize in [False, True]:
+    compiler = fhe.Compiler(f, {"x": "encrypted", "y": "encrypted", "z": "encrypted"})
+    circuit = compiler.compile(inputset, dataflow_parallelize=dataflow_parallelize)
+
+    circuit.keygen()
+    for sample in inputset[:3]:  # warmup
+        circuit.encrypt_run_decrypt(*sample)
+
+    timings = []
+    for sample in inputset[3:13]:
+        start = time.time()
+        result = circuit.encrypt_run_decrypt(*sample)
+        end = time.time()
+
+        assert np.array_equal(result, f(*sample))
+        timings.append(end - start)
+
+    if not dataflow_parallelize:
+        print(f"without dataflow parallelize -> {np.mean(timings):.03f}s")
+    else:
+        print(f"   with dataflow parallelize -> {np.mean(timings):.03f}s")
+```
+
+prints:
+
+```
+without dataflow parallelize -> 0.609s
+   with dataflow parallelize -> 0.414s
+```
+
+and the reason for that is:
+
+```
+// this is the generated MLIR for the circuit
+// without dataflow, every single line would be executed one after the other
```

Comment:
I wouldn't have the MLIR here, it's already too long

Reply:
I think it's valuable to explain how it works. Wdyt @BourgerieQuentin?